### ISSUE TITLE
unwrapping prevents composition

### ISSUE THREAD

**Issue by acthp** [jstat/jstat#168]
Various jStat methods, like multiply, dot, and transpose, unwrap matrices to vectors, and vectors to scalars, if they have length one. This prevents composing these methods into expressions. Instead, after every computation one has to check for these special cases and re-wrap the values in the proper type if they've been demoted.

What is the idea here? Is there some way to avoid this?

**Comment by trevnorris** [comment]
Initial idea was to simply return what was produced. For improved composability should jStat coerce values into matrices? I'd be okay w/ returning all vectors as matrices, but I think not wrapping scalars is the better option there.

**Comment by acthp** [comment]
The issue is that in an expression like

    (a, b) => multiply(multiply(b, inv(a)), transpose(b))

if the inner multiply results in a 1x1 matrix it will be demoted to a scalar, which then throws an error in the outer multiply, since the 1st argument to multiply should be a matrix.

So the extra code in transpose() and multiply() that changes the returned data type requires writing extra code in the callers to change the data type back. I think that code should just be dropped: return the matrix results as they are.

**Comment by acthp** [comment]
Specifically, I'm suggesting changing this line:

https://github.com/jstat/jstat/blob/master/src/linearalgebra.js#L60

to just do ```return res```. And this line:

https://github.com/jstat/jstat/blob/master/src/core.js#L201

to do ```return obj```. There are a few others that probably cause the same problem, like ```map``` and ```cumreduce```.

**Comment by trevnorris** [comment]
Hm. In retrospect that seems better. If you'd like feel free to open a PR with all these changes.

**Comment by jamescgibson** [comment]
:+1: on this idea - happy to do the PR as well. Though, this might break compatibility - would this need to go into a major version release?

**Comment by trevnorris** [comment]
@jamescgibson Yeah, it'll be in a major release, but since we don't have any strict release schedule that can happen whenever. If you make the PR i'll cut a branch for the current major and we can land it. Then can maintain the old major as necessary.
